<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<html>
    <head>
        <title>About VAMP_RNA_Trimming</title>
        <link rel="stylesheet" href="nbdocs://org.netbeans.modules.usersguide/org/netbeans/modules/usersguide/ide.css" type="text/css">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
        <style type="text/css">
            .mono {
                font-family: monospace;
            }
        </style>
    </head>
    <body>       
        <h2>Trimming</h2>
        <p> 
            With the trimming function a SAM file can be parsed for unmapped
            reads. In this process the reads will be shortened by using the 
            chosen trimming method. 
        </p>
        <p>
            This component is based on regular exception. There are currently
            6 implemented methods, which are available to the user for selection. 
            The following table shows the implemented methods and the 
            result of the given method used on the following example read: 
            <span class="mono">AATGCGGTATATAAAAA</span>. 
            The generation of the regular expression used in the program 
            takes place in dependence of maximum length parameter. 
            When using the trimming from both sides (3' and 5') this value is halved, 
            and the appropriate number of nucleotides on each side is shortened. 
            For the same reason, the value range is limited to just even numbers.
        </p>
        <table>
            <thead>
                <tr>
                    <th>Method name</th>
                    <th>Regular expression</th>
                    <th>Example effect</th>
                </tr>    
            </thead>
            <tbody>
                <tr>
              <td>trim poly-A from 3' end (right to left) by variable length</td>
              <td><span class="mono">^(.*?)(A{0,4})$</span></td>
              <td><span class="mono">AATGCGGTATATA----</span></td>
            </tr>
            <tr>
              <td>trim poly-A from 5' end (left to right) by variable length</td>
              <td><span class="mono">^(A{0,4})(.*?)$</span></td>
              <td><span class="mono">--TGCGGTATATAAAAA</span></td>
            </tr>
            <tr>
              <td>trim poly-A from 3' and from 5' end by variable length</td>
              <td><span class="mono">^(A{0,2})(.*?)(A{0,2})$</span></td>
              <td><span class="mono">--TGCGGTATATAAA--</span></td>
            </tr>
            <tr>
              <td>trim all nucleotides from 3' end (right to left) by fixed length</td>
              <td><span class="mono">^(.*?)(.{4})$</span></td>
              <td><span class="mono">AATGCGGTATATA----</span></td>
            </tr>
            <tr>
              <td>trim all nucleotides from 5' end (left to right) by fixed length</td>
              <td><span class="mono">^(.{4})(.*?)$</span></td>
              <td><span class="mono">----CGGTATATAAAAA</span></td>
            </tr>
            <tr>
              <td>trim all nucleotides from 3' and from 5' end by fixed length</td>
              <td><span class="mono">^(.{2})(.*?)(.{2})$</span></td>
              <td><span class="mono">--TGCGGTATATAAA--</span></td>
            </tr>
            </tbody>
        </table>
        <h3>Trim parameters</h3>
        <p> 
            <b>Reference FASTA file:</b>
            this reference sequence will be used for the mapping of 
            the trimmed sequences
        </p>
        <p> 
            <b>Mapped SAM Souce File:</b>
            provide a file which contains some unmapped reads. Only unmapped 
            reads will be extracted in a new FNA file, trimmed and mapped against 
            your reference sequence again. 
        </p>
        <p> 
            <b>Mapping parameter:</b>
            you can provide additional mapping parameters for the used mapping script.
        </p>
        <p> 
            <b>Maximum nucleotides to trim:</b>
            variable length trim methods will use this length as the maximum 
            number of nucleotides, that will be deleted from the sequence. 
            If you use fixed length trim methods, then this parameter will determine
            the exact number of trimmed nucleotides. If you use two-sided trim methods,
            the given number is used as the upper bound for the sum of the trimmed
            nucleotide count. This means, that only the half of this number is shortened
            on each side.
        </p>
    </body>
</html>
<!--
Tip: to create a link which will open in an external web browser, try:
<object classid="java:org.netbeans.modules.javahelp.BrowserDisplayer">
    <param name="content" value="http://www.netbeans.org/">
    <param name="text" value="<html><u>http://www.netbeans.org/</u></html>">
    <param name="textFontSize" value="medium">
    <param name="textColor" value="blue">
</object>
To create a link to a help set from another module, you need to know the code name base and path, e.g.:
<a href="nbdocs://org.netbeans.modules.usersguide/org/netbeans/modules/usersguide/configure/configure_options.html">Using the Options Window</a>
(This link will behave sanely if that module is disabled or missing.)
-->
