<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<html>
    <head>
        <title>About Transcription Analyses</title>
        <link rel="stylesheet" href="nbdocs:/org/netbeans/modules/usersguide/ide.css" type="text/css">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>
        <h2>About Transcription Analyses</h2>
        <p style="font-size:0.9em">
            The transcription analyses module enables you to carry out different automated
            analyses for transcriptome data sets. <br><br>
            1. The <b>Transcription Start Site (TSS) detection</b> (enabled by checking the "TSS detection" box) has two default parameters. 
            <br>A) The <u>minimal coverage increase</u> (absolute read counts) from one position to the neighboring one
            (default = 50 reads).<br> B) The <u>minimal coverage increase in percent</u> from one position 
            to the neighboring one (default = 50%). <!-- TODO: add help page with details on models-->
            <br><br>Additionally, two more parameters can be set: 
            <br> C) The <u>maximal initial read count</u> (default = 8 reads) can be set
            <br> D) The <u>minimal coverage increase</u> (default = 23%) for predicted TSS with less 
            than x (in the default case 8) reads at the position directly in front of the TSS has to be set. It can be the same as 
            the mandatory first parameter, but it is recommended to choose a lower value, depending on your analysis and data quality. 
            <br><br>If one of these two additional paramters is > 0 the other one also has to be set correctly.<br>
            By checking the <u>Automatic parameter estimation</u> box, the parameters are estimated from the data in the track. In this
            case the additional parameters are currently not set, only the mandatory parameters.<br><br>
            2. The <b>Filter Genes Analysis</b> allows to filter annotated genes of the reference sequence according to their raw read count.
            Therefore, it requires 
            <br> A) The <u>minimum number of reads</u> (default = 100 reads)
            <br> B) The <u>maximum number of reads</u> (default = 100000 reads) 
            <br>mapped within the bounds of an annotation.<br><br>
            3. The <b>Operon detection</b> allows to predict operons for prokaryotic genomes on the basis of:
            <br> A) <u>Minimal number of spanning reads</u> (default = 5 reads) 
            <br>between two annotated genes on the same strand. Reads are called "spanning", if they overlap both neighboring annotations.
            <br>For all these analyses the results are returned in a matrix. Furthermore, they are visualized in the corresponding track
            viewer by different colors. This helps to compare the different analysis results at a glance.
        </p>
    </body>
</html>
<!--
Tip: to create a link which will open in an external web browser, try:
<object classid="java:org.netbeans.modules.javahelp.BrowserDisplayer">
    <param name="content" value="http://www.netbeans.org/">
    <param name="text" value="<html><u>http://www.netbeans.org/</u></html>">
    <param name="textFontSize" value="medium">
    <param name="textColor" value="blue">
</object>
To create a link to a help set from another module, you need to know the code name base and path, e.g.:
<a href="nbdocs://org.netbeans.modules.usersguide/org/netbeans/modules/usersguide/configure/configure_options.html">Using the Options Window</a>
(This link will behave sanely if that module is disabled or missing.)
-->
