<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<html>
    <head>
        <title>About Transcription Analyses</title>
        <link rel="stylesheet" href="nbdocs:/org/netbeans/modules/usersguide/ide.css" type="text/css">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>
        <h2>About Transcription Analyses</h2>
        <p style="font-size:0.9em">
            The transcription analyses module enables you to carry out different automated
            analyses for transcriptome data sets. <br><br>
            1. The <b>Transcription Start Site (TSS) detection</b> (enabled by checking the "TSS detection" box in the wizard) has two default parameters. 
            <br>A) The <u>minimal number of read starts</u> at the examined position (default = 50 reads starts).
            <br>B) The <u>minimal coverage increase in percent</u> from the respective neighboring position 
            to the examined one (default = 50%). <!-- TODO: add help page with details on models-->
            <br><br>In order to output novel transcript information along with the detected TSS, you can choose the
            "<b>Novel transcript detection</b>". For detected TSS which do not have neighboring features, it adds the information,
            that this is a novel transcript and further a transcript stop is inferred by the coverage threshold <b>Minimal transcript 
            extension coverage</b>, which has to be set, if the novel transcript detection checkbox is selected.
            <br><br>Additionally, two more parameters can be set: 
            <br> C) The <u>maximal low coverage read start number</u> (default = 20 read starts)
            <br> D) The <u>minimal low coverage read start number</u> (default = 10 read starts).
            <br>For positions with less than x (in the default case 20) read starts at the examined position, but at least y
            (in the default case 10) read starts, which also fulfill parameter B), a TSS is predicted. It can be the same as 
            the mandatory first parameter, but it is recommended to choose a lower value, depending on your analysis and data quality. 
            <br><br>If one of these two additional paramters is > 0 the other one also has to be set correctly.<br>
            By checking the <u>Automatic parameter estimation</u> box, the parameters are estimated from the data in the track. In this
            case the additional parameters are not set, only the mandatory parameters.<br><br>
            2. The <b>RPKM and Read Count Calculation</b> calculates RPKM and raw read count values for annotated features of 
            the reference sequence. Therefore, it requires 
            <br> A) The <u>minimum number of reads</u> (default = 1 reads)
            <br> B) The <u>maximum number of reads</u> (default = 100000000 reads)
            <br>mapped within the bounds of an feature.<br><br>
            3. The <b>Operon detection</b> allows to predict operons for prokaryotic genomes on the basis of:
            <br> A) <u>Minimal number of spanning reads</u> (default = 5 reads) 
            <br>between two annotated genes on the same strand. Reads are called "spanning", if they overlap both neighboring features.
            <br>For all these analyses the results are returned in a matrix. Furthermore, they are visualized in the corresponding track
            viewer by different colors. This helps to compare the different analysis results at a glance.
        </p>
    </body>
</html>
<!--
Tip: to create a link which will open in an external web browser, try:
<object classid="java:org.netbeans.modules.javahelp.BrowserDisplayer">
    <param name="content" value="http://www.netbeans.org/">
    <param name="text" value="<html><u>http://www.netbeans.org/</u></html>">
    <param name="textFontSize" value="medium">
    <param name="textColor" value="blue">
</object>
To create a link to a help set from another module, you need to know the code name base and path, e.g.:
<a href="nbdocs://org.netbeans.modules.usersguide/org/netbeans/modules/usersguide/configure/configure_options.html">Using the Options Window</a>
(This link will behave sanely if that module is disabled or missing.)
-->
