<!--
To change this template, choose Tools | Templates
and open the template in the editor.
-->
<html>
    <head>
        <title>Import Data</title>
        <link rel="stylesheet" href="nbdocs:/org/netbeans/modules/usersguide/ide.css" type="text/css">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    </head>
    <body>
        <p style="text-align: center;">&nbsp;<span style="font-size: 14pt;"><strong>Import Data:</strong></span></p>
        <ol>
            <li style="text-align: justify;">Either choose the respective toolbar icon (see <i><a href="nbdocs://de.cebitec.readxplorer.ui/de/cebitec/readxplorer/ui/docs/view-about.html">Main Window a)</a></i>) or go to "<span style="text-decoration: underline;">File -&gt; Import data</span>".</li>
            <li style="text-align: justify;">Choose the tab of the data type you want to import. Note that at first a reference has to be imported and then tracks can be imported on that reference.</li>
            <li style="text-align: justify;">If a reference or track, which has been added to the import list, shall not be imported, just select in the list and click "<span style="text-decoration: underline;">Remove</span>" in the respective tab.</li>
        </ol>
        <p style="text-align: justify;"><span style="text-decoration: underline;"><strong>Import a reference sequence/genome:</strong></span></p>
        <ol>
            <li style="text-align: justify;"><strong>Make sure that all references have the same locus name as used by the reference file during the mapping process of the tracks, which shall be imported afterwards.</strong></li>

            <li style="text-align: justify;">Select the "References" tab and click "<span style="text-decoration: underline;">Add</span>".</li>
            <li style="text-align: justify;">Select the file type to import: Genbank, EMBL, GFF3, GFF2 or plain FASTA</li>
            <li style="text-align: justify;">Choose a file of the selected file type</li>
            <li style="text-align: justify;">Choose a name for the reference.</li>
            <li style="text-align: justify;">Enter a description of the reference, by default, it is the file name.</li>
            <li style="text-align: justify;">Click "<span><span style="text-decoration: underline;">OK</span></span>" to add the reference to the import list.</li>
        </ol>
        <p style="text-align: justify;"><strong><span style="text-decoration: underline;"><span lang="EN-US">Import mapping data sets:</span></span></strong></p>
        <p style="text-align: justify;" >When we talk about "importing" track data (= read mapping data set) here, it means that ReadXplorer reads the corresponding SAM/BAM or SARUMAN (JOK) input file and adds read classification information (see <i><a href="nbdocs://de.cebitec.readxplorer.ui/de/cebitec/readxplorer/ui/docs/classification-about.html">Read Classification</a></i> and <i><a href="nbdocs://de.cebitec.readxplorer.ui/de/cebitec/readxplorer/ui/docs/classification-about.html">Read Pair Classification</a></i>) into a new "copy" of the original input file. The copy contains all original information plus the read classification, carried out by ReadXplorer. <span style="text-decoration: underline;">During the import process your original SAM/BAM file is never altered.</span> ReadXplorer creates intermediate files during the import process next to the original SAM/BAM file. The temporary directory for these files can be configured via <span style="text-decoration: underline;">"Tools->Options->Miscellaneous"</span> (see <i><a href="nbdocs://de.cebitec.readxplorer.ui.options/de/cebitec/readxplorer/ui/options/docs/options-about.html">General Options</a></i>). At the end an extended version of the original file is created in the same directory. Besides the mapping, an index file is created for quicker access of the data.
            <br />When a track is opened, all data is read directly from the corresponding extended mapping file. If this file is moved, then ReadXplorer asks for the new path to that mapping file. If the index file should be missing, ReadXplorer creates a new index file.</p>
        <p style="text-align: justify;"><strong><span style="text-decoration: underline;"><span lang="EN-US">Mismatching reference identifiers in mappings and reference:</span></span></strong></p>        
        <p style="text-align: justify;" >Since version 2.2, ReadXplorer supports automatic and manual correction of mismatching reference identifiers between mapping files and the chosen reference. This can happen for example when downloading the Genbank and Fasta reference files from NCBI. Both files have a different locus tag and sequence identifier respectively. When the identifiers in the mapping file are not adapted, no data will be shown after importing it into ReadXplorer â€“ simply because no data is found for the reference chromosome. The mapping is typically performed on the Fasta file and for downstream analysis in ReadXplorer the Genbank reference is used to have the genome annotations at hand. 
            To avoid any problems, ALWAYS MAKE SURE THAT BOTH REFERENCE FILES (the annotated and the Fasta file) CONTAIN THE SAME SEQUENCE IDENTIFIERS.
            If this is not the case for some reason, the procedure for automatically correcting the sequence identifiers is as follows:</p>
        <ol>
            <li style="text-align: justify;">The REFERENCE SEQUENCE IDENTIFIERS are NEVER ALTERED, since this would also affect other tracks imported for that reference.</li>
            <li style="text-align: justify;">For references with only one chromosome, the identifier from the reference is written into the BAM mapping file.</li>
            <li style="text-align: justify;">For the mapping file ids, several special characters (:,/,\,*,?,|,<,>,") are replaced by '-'.</li>
            <li style="text-align: justify;">For the mapping file ids, a check is performed if replacement of "chr_x", "chrx", "chromosome_x" or "chromosomex" with each other leads to a match to one of the reference identifiers.</li>
            <li style="text-align: justify;">Finally, "chr" or "chromosome" prefixes are removed and only the chromosome number is retained and checked against the reference identifiers.</li>
        </ol>
        <p style="text-align: justify;" >When the automatic correction still fails for some of the references identifiers, a wizard for manually correction opens automatically (see below). When multiple files are imported, all mapping files containing exactly the same sequence identifiers are handled on a single wizard page. If their sequence dictionary is different, a new wizard page is added for each mapping file with a unique dictionary.</p>
        <p style="text-align: center;"><img style="display: block; margin-left: auto; margin-right: auto;" src="nbdocs://de.cebitec.readxplorer.ui.importer/de/cebitec/readxplorer/ui/importer/docs/mapping-id-correction.png" alt="mapping id correction figure" /></p>
        <p style="text-align: justify;" >The wizard lists all identifiers present in the reference and in the mapping file(s) and enables replacement of ids of the mapping file (right column) by ids present in the reference (left column) via drag and drop. Note that the order of mapping file ids cannot be altered, since this would compromise the assignment between mappings and their ids. Multiple columns can be dragged at once.</p>

        <p style="text-align: justify;"><strong><span style="text-decoration: underline;">Import one or more single end track(s):</span></strong></p>
        <p style="text-align: center;"><img style="display: block; margin-left: auto; margin-right: auto;" src="nbdocs://de.cebitec.readxplorer.ui.importer/de/cebitec/readxplorer/ui/importer/docs/import-singleendtrack.jpg" alt="TrackImport" /></p>
        <ol>
            <li style="text-align: justify;">Select the "<span style="text-decoration: underline;">Tracks</span>" tab and click "<span style="text-decoration: underline;">Add</span>".<br /></li>
            <li style="text-align: justify;">Select the reference genome, for which the single end track shall be imported.</li>
            <li style="text-align: justify;">Select the mapping parser to use: For SAM/BAM files use the "<span style="text-decoration: underline;">SAM/BAM Direct Access Parser</span>" and for SARUMAN mapping output (JOK) use the "<span style="text-decoration: underline;">Jok to BAM Direct Access Parser</span>".</li>
            <li style="text-align: justify;">Click "<span style="text-decoration: underline;">Open</span>" to select one or more mapping files to import. When importing multiple files, they have to be contained in the same directory. Chosen multiple files are then listed in the "<span style="text-decoration: underline;">Import track list</span>".</li>
            <li style="text-align: justify;">Change the description for a single track import.</li>
            <li style="text-align: justify;">If the specified track(s) have already been used in another ReadXplorer DB, check the box "<span style="text-decoration: underline;">Was already imported in another ReadXplorer DB</span>".In this case the read classification is already stored in the bam file and does not need to be recalculated. This speeds up the import process.</li>
            <li style="text-align: justify;">Click "<span style="text-decoration: underline;">OK</span>" to add the track(s) to the import list.</li>
            <li style="text-align: justify;">If more tracks from another directory shall be imported, start with 1. again.</li>
        </ol>
        <p style="text-align: justify;"><strong><span style="text-decoration: underline;">Import one or more paired end or mate pair track(s):</span></strong>
        <p style="text-align: center;"><img style="display: block; margin-left: auto; margin-right: auto;" src="nbdocs://de.cebitec.readxplorer.ui.importer/de/cebitec/readxplorer/ui/importer/docs/import-readpairs.jpg" alt="import-readpairs" /></p>
        <ol>
            <li style="text-align: justify;">Select the "<span style="text-decoration: underline;">Read Pair Tracks</span>" tab and click "<span style="text-decoration: underline;">Add</span>".<br style="clear: none;" /> </li>
            <li style="text-align: justify;">Select the mapping parser to use: For SAM/BAM files use the "<span style="text-decoration: underline;">SAM/BAM Direct Access Parser</span>" and for SARUMAN mapping output (JOK) use the "<span style="text-decoration: underline;">Jok to BAM Direct Access Parser</span>".</li>
            <li style="text-align: justify;">Click "<span style="text-decoration: underline;">Open</span>" to select one or more mapping files to import. Since read pair mappings are sometimes contained in two separate files (for read 1 and read 2), it is possible to add two mapping files. But only the "<span style="text-decoration: underline;">Mapping file read 1</span>" has to be selected. The "<span style="text-decoration: underline;">Mapping file read 2</span>" can be blank if all reads are contained in one file. When importing multiple files, they have to be contained in the same directory. The multiple files are then listed in the "<span style="text-decoration: underline;">Import track list</span>". <strong>Multiple files for read 1 and read 2 have to be in the same alphabetical order in order to be combined.</strong> </li>
            <li style="text-align: justify;">Change the description for a single read pair track import.</li>
            <li style="text-align: justify;">Select the "<span style="text-decoration: underline;">Complete distance</span>", which is the expected distance from the start of a read pair to the end of the pair (e.g. read 1 (100-150), read 2 (300-350) has distance 250) of the read pairs.</li>
            <li style="text-align: justify;">Select the maximum allowed "<span style="text-decoration: underline;">% deviation</span>" from the "<span style="text-decoration: underline;">Complete distance</span>" to classify a read pair as perfect.</li>
            <li style="text-align: justify;">Select the direction of the read pairs: forward-reverse (fr) for paired-end, reverse-forward (rf) for mate pair or forward-forward/reverse-reverse (ff/rr) for other libraries.</li>
            <li style="text-align: justify;">Select the reference genome, for which the read pair track shall be imported.</li>
            <li style="text-align: justify;">If the specified read pair track(s) have already been used in another ReadXplorer DB as read pair tracks, check the box "<span style="text-decoration: underline;">Already imported as READ PAIR track into another ReadXplorer DB</span>".</li>
            <li style="text-align: justify;">Click "<span style="text-decoration: underline;">OK</span>" to add the track(s) to the import list.</li>
            <li style="text-align: justify;">If more read pair tracks from another directory shall be imported, start with 1. again.<br /></li>
        </ol>

    </body>
</html>
<!--
Tip: to create a link which will open in an external web browser, try:
<object classid="java:org.netbeans.modules.javahelp.BrowserDisplayer">
    <param name="content" value="http://www.netbeans.org/">
    <param name="text" value="<html><u>http://www.netbeans.org/</u></html>">
    <param name="textFontSize" value="medium">
    <param name="textColor" value="blue">
</object>
To create a link to a help set from another module, you need to know the code name base and path, e.g.:
<a href="nbdocs://org.netbeans.modules.usersguide/org/netbeans/modules/usersguide/configure/configure_options.html">Using the Options Window</a>
(This link will behave sanely if that module is disabled or missing.)
-->
