package de.cebitec.readXplorer.transcriptomeAnalyses.mainWizard;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.BorderFactory;
import javax.swing.JCheckBox;
import javax.swing.JPanel;

/**
 *
 * @author jritter
 */
public final class DataSetChoiceVisualPanel extends JPanel implements ActionListener {

    private final String fiveprimeText = "Your cDNA Libraries were prepared with a protocol comparable or similar\n"
            + " to the one described in Pfeifer-Sancar et al., 2013 and single read sequencing was performed.\n"
            + " Here, it is of importance that primary transcripts were enriched with a suitable method such \n"
            + "as Terminator exonuclease treatment.";
    private final String wholeTranscriptText = "Your cDNA libraries were prepared with a protocol comparable \n "
            + "or similar to the one described in Pfeifer-Sancar et al., 2013.\n"
            + " and paired end sequencing was performed. To ensure that calculations work properly,\n"
            + " it is necessary that the gap between paired reads is extended in silico (cf. figure).";

    /**
     * Creates new form DataSetChoiceVisualPanel
     */
    public DataSetChoiceVisualPanel() {
        initComponents();

        wholeTranscriptomeTrackCheckBox.addActionListener(this);
        wholeTranscriptomeTrackCheckBox.setActionCommand(TranscriptomeAnalysisWizardIterator.PROP_WHOLEGENOME_DATASET);
        fiveEnrichedTrackCheckBox.addActionListener(this);
        fiveEnrichedTrackCheckBox.setActionCommand(TranscriptomeAnalysisWizardIterator.PROP_FIVEPRIME_DATASET);

        wholeTranscriptTA.setBorder(BorderFactory.createTitledBorder("Whole transcript RNA-Seq data set"));
        wholeTranscriptTA.setText(wholeTranscriptText);
        fivePrimeExplanationTA.setBorder(BorderFactory.createTitledBorder("5'-Enriched RNA-Seq data set"));
        fivePrimeExplanationTA.setText(fiveprimeText);
    }

    @Override
    public String getName() {
        return "Data selection";
    }

    public boolean isWholeGenomeTrack() {
        return this.wholeTranscriptomeTrackCheckBox.isSelected();
    }

    public boolean isFiveEnrichedTrack() {
        return this.fiveEnrichedTrackCheckBox.isSelected();
    }

    /**
     * This method is called from within the constructor to initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is always
     * regenerated by the Form Editor.
     */
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        wholeTranscriptomeTrackCheckBox = new javax.swing.JCheckBox();
        fiveEnrichedTrackCheckBox = new javax.swing.JCheckBox();
        jLabel1 = new javax.swing.JLabel();
        jSeparator1 = new javax.swing.JSeparator();
        jScrollPane1 = new javax.swing.JScrollPane();
        wholeTranscriptTA = new javax.swing.JTextArea();
        jScrollPane2 = new javax.swing.JScrollPane();
        fivePrimeExplanationTA = new javax.swing.JTextArea();

        org.openide.awt.Mnemonics.setLocalizedText(wholeTranscriptomeTrackCheckBox, org.openide.util.NbBundle.getMessage(DataSetChoiceVisualPanel.class, "DataSetChoiceVisualPanel.wholeTranscriptomeTrackCheckBox.text_1")); // NOI18N

        org.openide.awt.Mnemonics.setLocalizedText(fiveEnrichedTrackCheckBox, org.openide.util.NbBundle.getMessage(DataSetChoiceVisualPanel.class, "DataSetChoiceVisualPanel.fiveEnrichedTrackCheckBox.text")); // NOI18N
        fiveEnrichedTrackCheckBox.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                fiveEnrichedTrackCheckBoxActionPerformed(evt);
            }
        });

        org.openide.awt.Mnemonics.setLocalizedText(jLabel1, org.openide.util.NbBundle.getMessage(DataSetChoiceVisualPanel.class, "DataSetChoiceVisualPanel.jLabel1.text")); // NOI18N

        wholeTranscriptTA.setColumns(20);
        wholeTranscriptTA.setRows(5);
        jScrollPane1.setViewportView(wholeTranscriptTA);

        fivePrimeExplanationTA.setColumns(20);
        fivePrimeExplanationTA.setRows(5);
        jScrollPane2.setViewportView(fivePrimeExplanationTA);

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(jLabel1, javax.swing.GroupLayout.DEFAULT_SIZE, 645, Short.MAX_VALUE)
                    .addComponent(jSeparator1)
                    .addComponent(jScrollPane1)
                    .addGroup(layout.createSequentialGroup()
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addComponent(fiveEnrichedTrackCheckBox)
                            .addComponent(wholeTranscriptomeTrackCheckBox))
                        .addGap(0, 0, Short.MAX_VALUE))
                    .addComponent(jScrollPane2))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(jLabel1, javax.swing.GroupLayout.PREFERRED_SIZE, 30, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jSeparator1, javax.swing.GroupLayout.PREFERRED_SIZE, 10, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                .addComponent(wholeTranscriptomeTrackCheckBox)
                .addGap(13, 13, 13)
                .addComponent(jScrollPane1, javax.swing.GroupLayout.PREFERRED_SIZE, 80, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(fiveEnrichedTrackCheckBox)
                .addGap(18, 18, 18)
                .addComponent(jScrollPane2, javax.swing.GroupLayout.PREFERRED_SIZE, 80, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );
    }// </editor-fold>//GEN-END:initComponents

    private void fiveEnrichedTrackCheckBoxActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_fiveEnrichedTrackCheckBoxActionPerformed
        // TODO add your handling code here:
    }//GEN-LAST:event_fiveEnrichedTrackCheckBoxActionPerformed
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JCheckBox fiveEnrichedTrackCheckBox;
    private javax.swing.JTextArea fivePrimeExplanationTA;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JScrollPane jScrollPane1;
    private javax.swing.JScrollPane jScrollPane2;
    private javax.swing.JSeparator jSeparator1;
    private javax.swing.JTextArea wholeTranscriptTA;
    private javax.swing.JCheckBox wholeTranscriptomeTrackCheckBox;
    // End of variables declaration//GEN-END:variables

    @Override
    public void actionPerformed(ActionEvent e) {
        if (e.getActionCommand().equals(TranscriptomeAnalysisWizardIterator.PROP_WHOLEGENOME_DATASET)) {
            if (((JCheckBox) e.getSource()).isSelected()) {
                fiveEnrichedTrackCheckBox.setSelected(false);
            }
        }

        if (e.getActionCommand().equals(TranscriptomeAnalysisWizardIterator.PROP_FIVEPRIME_DATASET)) {
            if (((JCheckBox) e.getSource()).isSelected()) {
                wholeTranscriptomeTrackCheckBox.setSelected(false);
            }
        }
    }
}
